ADR 033: Inter-Module Communication (#7459)
cosmos/cosmos-sdk: docs/architecture/adr-033-protobuf-inter-module-comm.md

# ADR 033: Protobuf-based Inter-Module Communication

## Changelog

- 2020-10-05: Initial Draft

## Status

Proposed

## Abstract

This ADR introduces a system for permissioned inter-module communication leveraging the protobuf `Query` and `Msg`
service definitions defined in [ADR 021](./adr-021-protobuf-query-encoding.md) and
[ADR 031](./adr-031-msg-service.md) which provides:
- stable module interfaces to eventually replace the keeper paradigm based on protobuf
- stronger inter-module object capabilities guarantees
- module accounts and sub-account authorization

## Context

In the current Cosmos SDK documentation on the [Object-Capability Model](../docs/core/ocap.md), it is state that:

> We assume that a thriving ecosystem of Cosmos-SDK modules that are easy to compose into a blockchain application will contain faulty or malicious modules.

There is currently not a thriving ecosystem of Cosmos SDK modules. We hypothesize that this is in part due to:

> 1. lack of a stable v1.0 Cosmos SDK to build modules off of. Module interfaces are changing, sometimes dramatically, from

Reviewer: I agree, the lack of a stable v1.0 is becoming an increasingly bigger problem. Though it's unclear to me how the design outlined here could help migrate the Cosmos SDK to achieve stability? Isn't this just another way to workaround such longstanding issue?

Author: Well, for one buf break check could help. But sure you could have some other way of doing that with plain go. I think the big thing that it helps with is enforcing proper isolation. There are a lot of methods in say the `x/auth` and `x/bank` `Keeper` interfaces that really shouldn't be exposed that generally. So a lot of the benefit here I see is around enforcing proper authorization or "Ocaps"

Reviewer: > But sure you could have some other way of doing that with plain go.

Yeah, correct ! That's my point.

> I think the big thing that it helps with is enforcing proper isolation.

But you can have that by strinking the right balance between public packages and `internal/`, can't you?

Author: > But you can have that by strinking the right balance between public packages and internal/, can't you?

I mean, in theory that would have been done with existing keepers, but it wasn't and people would be enforcing backwards compatibility with plain go, but we're not. Sometimes, the guard rails need to be built into the protocol. This is a roadmap for that.

point release to point release, often for good reasons, but this does not create a stable foundation to build on.

> 2. lack of a properly implemented object capability or even object-oriented encapsulation system which makes refactors

Author: Well, I think the fact that IBC needs to prevent other modules from mess with their (`ibc/`) denoms one case: https://github.com/cosmos/cosmos-sdk/issues/7093#issuecomment-726675035. Basically because `x/bank` exposes so much, it makes a refactor constraining that interface inevitable IMHO.

> 2. lack of a properly implemented object capability or even object-oriented encapsulation system which makes refactors

Author: So starting from the point of view that inter-module calls should be authenticated would make that sort of refactoring unneeded. If everything were authenticated to start, we would never need to address it as a security concern later.

of module keeper interfaces inevitable because the current interfaces are poorly constrained.

### `x/bank` Case Study

> Currently the `x/bank` keeper gives pretty much unrestricted access to any module which references it. For instance, the

Author: As I'm pointing out, every module has access to `SetBalance` which even can break supply invariants. So what exactly is restricted?

> Currently the `x/bank` keeper gives pretty much unrestricted access to any module which references it. For instance, the

Reviewer 2: The interface provided the module. If the interface doesn't specify `SetBalance`, it cannot call `SetBalance`. So it is restricted to the interface required by the module.

> Currently the `x/bank` keeper gives pretty much unrestricted access to any module which references it. For instance, the

Author: Right, but modules can always cast to a module that has `SetBalance` if they know it's there. From an auditing perspective, it's easier to see that a method isn't publicly exposed vs an expected keeper not declaring it. And many expected keepers do declare methods like this.

> Currently the `x/bank` keeper gives pretty much unrestricted access to any module which references it. For instance, the

Reviewer 2: I mean, sure, but why would you ever do that.

### New "Keeper" Paradigm

In [ADR 021](./adr-021-protobuf-query-encoding.md), a mechanism for using protobuf service definitions to define queriers
was introduced and in [ADR 31](./adr-031-msg-service.md), a mechanism for using protobuf service to define `Msg`s was added.
Protobuf service definitions generate two golang interfaces representing the client and server sides of a service plus
some helper code. Here is a minimal example for the bank `Send` `Msg`.

```go
package bank

type MsgClient interface {
	Send(context.Context, *MsgSend, opts ...grpc.CallOption) (*MsgSendResponse, error)
}

type MsgServer interface {
	Send(context.Context, *MsgSend) (*MsgSendResponse, error)
}
```

[ADR 021](./adr-021-protobuf-query-encoding.md) and [ADR 31](./adr-031-msg-service.md) specifies how modules can implement the generated `QueryServer`
and `MsgServer` interfaces as replacements for the legacy queriers and `Msg` handlers respectively.

In this ADR we explain how modules can make queries and send `Msg`s to other modules using the generated `QueryClient`
and `MsgClient` interfaces and propose this mechanism as a replacement for the existing `Keeper` paradigm.

> Using this `QueryClient`/`MsgClient` approach has the following key benefits over keepers:

Author: Sure. They can still be called keepers. And we don't even need to replace existing APIs, but simply have a new more explicit pathway for API interfaces.

is needed. We introduce a new type, `ModuleKey`, to serve this role which we can conceptualize as the "private key"
corresponding to a module account.

Whereas external clients use their private key to sign transactions containing `Msg`s where they are listed as signers,
modules use their `ModuleKey` to send `Msg`s where they are listed as the sole signer to other modules. For example, modules
could use their `ModuleKey` to "sign" a `/cosmos.bank.Msg/Send` transaction to send coins from the module's account to another
account.

`QueryClient`s could also be made with `ModuleKey`s, except that authentication isn't required.

Here's an example of a hypothetical module `foo` interacting with `x/bank`:
```go
package foo

func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBar) (*MsgBarResponse, error) {
  bankQueryClient := bank.NewQueryClient(fooMsgServer.moduleKey)
  balance, err := bankQueryClient.Balance(&bank.QueryBalanceRequest{Address: fooMsgServer.moduleKey.Address(), Denom: "foo"})
  
  ...

  bankMsgClient := bank.NewMsgClient(fooMsgServer.moduleKey)

>   res, err := bankMsgClient.Balance(ctx, &bank.MsgSend{FromAddress: fooMsgServer.moduleKey.Address(), ...})

Reviewer 3: ```suggestion
  res, err := bankMsgClient.Send(ctx, &bank.MsgSend{FromAddress: fooMsgServer.moduleKey.Address(), ...})
```

  cfg.RequireServer((*bank.MsgServer)(nil))
}
```

### Security Considerations

In addition to checking for `ModuleKey` permissions, a few additional security precautions will need to be taken by
the underlying router infrastructure.

#### Recursion and Re-entry

Recursive or re-entrant method invocations pose a potential security threat. This can be a problem if Module A
calls Module B and Module B calls module A again in the same call.

One basic way for the router system to deal with this is to maintain a call stack which prevents a module from
being referenced more than once in the call stack so that there is no re-entry. A `map[string]interface{}` table
in the router could be used to perform this security check.

#### Queries

Queries in Cosmos SDK are generally un-permissioned so allowing one module to query another module should not pose
any major security threats assuming basic precautions are taken. The basic precaution that the router system will
need to take is making sure that the `sdk.Context` passed to query methods does not allow writing to the store. This
can be done for now with a `CacheMultiStore` as is currently done for `BaseApp` queries.

> ### Internal Methods

Reviewer 4: I'm still half-convinced that exposing an Internal service is the way to go.

One question about the hooks/plugins model (without going into its details): How will the hooks/plugins model work along with the Internal service model? Does one deprecate/supersede another?

A comparison using a concrete example (e.g. staking/slashing communication) would be ideal.

Author: I can start documenting my ideas hooks/plugins. Would it be okay if I do that in another draft PR?

Reviewer 4: Yes, that'd be ideal.

By default, the inter-module router requires that messages are sent by the first signer returned by `GetSigners`. The
inter-module router should also accept authorization middleware such as that provided by [ADR 030](https://github.com/cosmos/cosmos-sdk/pull/7105).
This middleware will allow accounts to otherwise specific module accounts to perform actions on their behalf.
Authorization middleware should take into account the need to grant certain modules effectively "admin" privileges to
other modules. This will be addressed in separate ADRs or updates to this ADR.

### Future Work

Other future improvements may include:
* custom code generation that:
  * simplifies interfaces (ex. generates code with `sdk.Context` instead of `context.Context`)
  * optimizes inter-module calls - for instance caching resolved methods after first invocation
* combining `StoreKey`s and `ModuleKey`s into a single interface so that modules have a single Ocaps handle
* code generation which makes inter-module communication more performant
* decoupling `ModuleKey` creation from `AppModuleBasic.Name()` so that app's can override root module account names
* inter-module hooks and plugins

## Alternatives

> The `x/capability` module does provide a proper object-capability implementation that can be used by any module in the

Reviewer 5: I think this approach is preferable given that the amount of changes is less than the proposed changes in this ADR.

SDK and could even be used for inter-module Ocaps as described in [\#5931](https://github.com/cosmos/cosmos-sdk/issues/5931).

The advantages of the approach described in this ADR are mostly around how it integrates with other parts of the SDK,
specifically:

* protobuf so that:
  * code generation of interfaces can be leveraged for a better dev UX
  * module interfaces are versioned and checked for breakage using [buf](https://docs.buf.build/breaking-overview)
* sub-module accounts as per ADR 028
* the general `Msg` passing paradigm and the way signers are specified by `GetSigners`

Also, this is a complete replacement for keepers and could be applied to _all_ inter-module communication whereas the
`x/capability` approach in #5931 would need to be applied method by method.

> 

Reviewer 6: Was a pure go implementation (no proto, not `x/capability`) considered? If so, can it be documented to why it was not chosen?

Author: See #7093. I can go into a more detailed comparison if that's helpful

Reviewer 6: oh thanks for this, helps explain the rationale. Can it be documented in the ADR?

## Consequences

### Backwards Compatibility

This ADR is intended to provide a pathway to a scenario where there is greater long term compatibility between modules.
In the short-term, this will likely result in breaking certain `Keeper` interfaces which are too permissive and/or

> replacing `Keeper` interfaces altogether.

Reviewer 5: this is a massive change even for apps that only use SDK modules like Gaia. Just can't imagine the burden for apps like Kava which have 10 custom modules

Author: So like I've said, there's a way we can _enable_ this without converting everything to use it. We don't need to just get rid of old keeper interfaces to add this approach. And it doesn't need to be an all other nothing thing. I'm planning to update the ADR to reflect this. Refactoring module by module is sort of separate discussion clearly.

### Positive

- an alternative to keepers which can more easily lead to stable inter-module interfaces
- proper inter-module Ocaps

### Negative

> - modules which adopt this will need significant refactoring

Reviewer 5: I'd prefer an alternative that doesn't cause this amount of refactoring. The dev UX gain from this ADR is probably not enough to make it worth for SDK devs to implement this solution. Stargate has a massive amount of changes, and I honestly think we should reconsider the amount of breaking changes that new ADRs generate.

Author: Commented on the alternative a few places, namely _converting_ is a different discussion from _enabling_.

Reviewer 7: I agree.... there is a huge exhaustion of following breaking changes *every sdk release* for 2 years now.

It is not just stargate, but even 0.34 -> 0.36 or 0.37 -> 0.38 upgrades were quite painful. With the whole layout of the modules restructured (and re-restructured) as the core devs had a new, inspiring design. The change-fatigue and desire for stability in the ecosystem should be taking into consideration.

There are also many projects on mainnet with many custom modules and only a few of them are rushing to stargate. The more backwards-incompatible changes, the more we risk splintering the ecosystem.

> 

Reviewer 6: Protobuf has not been everyone's favorite to work with, considering integrating Protobuf deeper into the stack is a possible negative. Not sure how others view this.

Author: Can you share more context @marbar3778? This ADR by the way doesn't add any additional protobuf on its own - it just allows existing protobuf to be used for more purposes.

Reviewer 8: Maybe we can "deeper reuse of protobuf generated code"? I don't think it's negative, sounds more like neutral or positive.

2. lack of a properly implemented object capability or even object-oriented encapsulation system which makes refactors
of module keeper interfaces inevitable because the current interfaces are poorly constrained.

### `x/bank` Case Study

We use `x/bank` of this.

Currently the `x/bank` keeper gives pretty much unrestricted access to any module which references it. For instance, the
`SetBalance` method allows the caller to set the balance of any account to anything, bypassing even proper tracking of supply.

There appears to have been some later attempts to implement some semblance of Ocaps using module-level minting, staking
and burning permissions. These permissions allow a module to mint, burn or delegate tokens with reference to the module’s
own account. These permissions are actually stored as a `[]string` array on the `ModuleAccount` type in state.

However, these permissions don’t really do much. They control what modules can be referenced in the `MintCoins`,
`BurnCoins` and `DelegateCoins***` methods, but for one there is no unique object capability token that controls access

> - just a simple string. So the `x/upgrade` module could mint tokens for the `x/staking` module simple by calling

Reviewer 4: this dash here creates a new line with a bullet item

and `MsgClient` interfaces and propose this mechanism as a replacement for the existing `Keeper` paradigm.

Using this `QueryClient`/`MsgClient` approach has the following key benefits over keepers:
1. Protobuf types are checked for breaking changes using [buf](https://buf.build/docs/breaking-overview) and because of
the way protobuf is designed this will give us strong backwards compatibility guarantees while allowing for forward
evolution.
2. The separation between the client and server interfaces will allow us to insert permission checking code in between
the two which checks if one module is authorized to send the specified `Msg` to the other module providing a proper
object capability system.

This mechanism has the added benefits of:
- reducing boilerplate through code generation, and
- allowing for modules in other languages either via a VM like CosmWasm or sub-processes using gRPC 

### Inter-module Communication

In order for code to use the `Client` interfaces generated by the protobuf compiler, a `grpc.ClientConn` implementation
is needed. We introduce a new type, `ModuleKey`, to serve this role which we can conceptualize as the "private key"

> corresponding to a module account.

Reviewer 4: We could make it explicit that `ModuleKey` extends the `grpc.ClientConn` interface.

Whereas external clients use their private key to sign transactions containing `Msg`s where they are listed as signers,
modules use their `ModuleKey` to send `Msg`s where they are listed as the sole signer to other modules. For example, modules
could use their `ModuleKey` to "sign" a `/cosmos.bank.Msg/Send` transaction to send coins from the module's account to another
account.

> `QueryClient`s could also be made with `ModuleKey`s, except that authentication isn't required.

Author: Actually I need to remove this as we use the `ModuleKey` for queries too so clients don't need to decide which key to use.

Here's an example of a hypothetical module `foo` interacting with `x/bank`:
```go
package foo

func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBar) (*MsgBarResponse, error) {
  bankQueryClient := bank.NewQueryClient(fooMsgServer.moduleKey)
  balance, err := bankQueryClient.Balance(&bank.QueryBalanceRequest{Address: fooMsgServer.moduleKey.Address(), Denom: "foo"})
  
  ...

>   bankMsgClient := bank.NewMsgClient(fooMsgServer.moduleKey)

Reviewer 8: So how about creating required clients in the constructor and storing them as a private fields in the structure? It has few benefits:
* we clearly signal the dependencies
* we don't recreate the objects (caching).

  res, err := bankMsgClient.Balance(ctx, &bank.MsgSend{FromAddress: fooMsgServer.moduleKey.Address(), ...})

  ...
}
```

> ### `ModuleKey`s and `ModuleID`s

Reviewer 9: Is this `ModuleKey` intended to be used only for permissioning module accounts and sub-module accounts. Or could I use it to express arbitrary permissions?

For example, I want a supply module that provisions minting/burning permissions for different modules. So my supply module allows minting permission to ibc for any denomination that is prefixed by `ibc/` and nothing else, and no other module is allowed to mint denominations with `ibc/`.

That way I can ensure that no other module can mint `ibc/` tokens.

The current capability system is able to do things like this. I would prefer that we have a single dynamic inter-module capability system in the SDK. I have no strong opinion on whether it should be what is implemented in `x/capability` or this. But ideally we have a single system that supports all of our usecases, rather than multiple systems that support different subsets

Author: That could be accomplished with this system and is one of the goals laid out in #7093.

Author: The specifics of doing minting by denom prefix of course needs its own permission setup (as it would with either solution). But it's definitely a use case I'm holding in mind.

A `ModuleKey` can be thought of as a "private key" for a module account and a `ModuleID` can be thought of as the
corresponding "public key". From [the ADR 028 draft](https://github.com/cosmos/cosmos-sdk/pull/7086), modules can have both a root module account and any number of sub-accounts
or derived accounts that can be used for different pools (ex. staking pools) or managed accounts (ex. group
accounts). We can also think of module sub-accounts as similar to derived keys - there is a root key and then some
derivation path. `ModuleID` is a simple struct which contains the module name and optional "derivation" path,
and forms its address based on the `AddressHash` method from [the ADR 028 draft](https://github.com/cosmos/cosmos-sdk/pull/7086):

```go
type ModuleID struct {
  ModuleName string
  Path []byte
}

func (key ModuleID) Address() []byte {
  return AddressHash(key.ModuleName, key.Path)
}
```

> In addition to being able to generate a `ModuleID` and address, a `ModuleKey` contains a special function closure called

Reviewer 10: see `RootModuleKey` and `DerivedModuleKey` examples below, the `ModuleId` can be generated from the `moduleName` (and `path` in case of `DerivedModuleKey`), but I agree this would be great to make this clearer.

> In addition to being able to generate a `ModuleID` and address, a `ModuleKey` contains a special function closure called

Reviewer 10: Also, as outlined in this ADR, we can think of the `ModuleKey` as a private key and `ModuleId` as a public key.
So similar to our `PrivKey` interface which defines `PubKey()` method, `ModuleKey` defines a `ModuleId()` method.

the `Invoker` which is the key to safe inter-module access. This function closure corresponds to the `Invoke` method in
the `grpc.ClientConn` interface and under the hood is able to route messages to the appropriate `Msg` and `Query` handlers
performing appropriate security checks on `Msg`s. This allows for even safer inter-module access than keeper's whose
private member variables could be manipulated through reflection. Golang does not support reflection on a function
closure's captured variables and direct manipulation of memory would be needed for a truly malicious module to bypass
the `ModuleKey` security.

The two `ModuleKey` types are `RootModuleKey` and `DerivedModuleKey`:

```go

> func Invoker(callInfo CallInfo) func(ctx context.Context, request, response interface{}, opts ...interface{}) error

Reviewer 8: Syntax error. Should be:
```
type Invoker ...
```

Reviewer 8: ```suggestion
type Invoker(callInfo CallInfo) func(ctx context.Context, request, response interface{}, opts ...interface{}) error
```

type CallInfo {
  Method string
  Caller ModuleID
}

type RootModuleKey struct {
  moduleName string
  invoker Invoker

> }

Reviewer 8: Maybe we can make a `ModuleKey` an interface as well:
```
type ModuleKey {
    Name() string
    MkInvoke(callInfo CallInfo) grpc.Invoke  /// or InvokeFn or other appropriate type --> see comments above
    Derive(path []byte) ModuleKey
}
```

Reviewer 8: I think we also need a cryptographic key. Using only name for authentication is not secure - other modules could use the same name.

Author: That's not actually true if you study the underlying design. The name doesn't do any authentication, the function closure simply restricts access.

Reviewer 10: @robert-zaremba `ModuleKey` interface would rather look like:
```go
type ModuleKey interface {
	grpc.ClientConnInterface

	ModuleID() ModuleID
	Address() []byte
}
``` 
`grpc.ClientConnInterface` already defines an `Invoke` method.
`Derive()` method only makes sense for root `ModuleKey`s.

Reviewer 8: Yes, looks better. I made a draft based on what's in this ADR.

type DerivedModuleKey struct {
  moduleName string
  path []byte
  invoker Invoker
}
```

A module can get access to a `DerivedModuleKey`, using the `Derive(path []byte)` method on `RootModuleKey` and then
would use this key to authenticate `Msg`s from a sub-account. Ex:

```go
package foo

func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBar) (*MsgBarResponse, error) {
  derivedKey := fooMsgServer.moduleKey.Derive(req.SomePath)
  bankMsgClient := bank.NewMsgClient(derivedKey)
  res, err := bankMsgClient.Balance(ctx, &bank.MsgSend{FromAddress: derivedKey.Address(), ...})
  ...
}
```

In this way, a module can gain permissioned access to a root account and any number of sub-accounts and send
authenticated `Msg`s from these accounts. The `Invoker` `callInfo.Caller` parameter is used under the hood to
distinguish between different module accounts, but either way the function returned by `Invoker` only allows `Msg`s

> from either the root or a derived module account to pass through.

Reviewer 8: Is this secure (see my message above about cryptographic key)? How we make sure a module won't impersonalize other modules?

Author: Yes, the closure _only_ allows access to authorized calls. There is no access to anything else regardless of any name impersonation.

in the future that cache the invoke function for each method type avoiding the overhead of hash table lookup.
This would reduce the performance overhead of this inter-module communication method to the bare minimum required for
checking permissions.

Below is a rough sketch of the implementation of `grpc.ClientConn.Invoke` for `RootModuleKey`:

```go
func (key RootModuleKey) Invoke(ctx context.Context, method string, args, reply interface{}, opts ...grpc.CallOption) error {
  f := key.invoker(CallInfo {Method: method, Caller: ModuleID {ModuleName: key.moduleName}})
  return f(ctx, args, reply)
}
```

### `AppModule` Wiring and Requirements

In [ADR 031](./adr-031-msg-service.md), the `AppModule.RegisterService(Configurator)` method was introduced. To support
inter-module communication, we extend the `Configurator` interface to pass in the `ModuleKey` and to allow modules to
specify their dependencies on other modules using `RequireServer()`:

```go
type Configurator interface {
   MsgServer() grpc.Server
   QueryServer() grpc.Server

   ModuleKey() ModuleKey
   RequireServer(serverInterface interface{})

> }

Reviewer 8: If configurator is kind of a config, then it will be more natural to use a `struct` here, unless we require a Configurator to have certain methods and be able to mock them in a more advanced way. Basically I don't see why we prefer interface here - for simple mocking we can just set expected structure values.

Author: We are mocking it for testing in regen already. I think it's useful.

calls Module B and Module B calls module A again in the same call.

One basic way for the router system to deal with this is to maintain a call stack which prevents a module from
being referenced more than once in the call stack so that there is no re-entry. A `map[string]interface{}` table
in the router could be used to perform this security check.

#### Queries

Queries in Cosmos SDK are generally un-permissioned so allowing one module to query another module should not pose
any major security threats assuming basic precautions are taken. The basic precaution that the router system will
need to take is making sure that the `sdk.Context` passed to query methods does not allow writing to the store. This
can be done for now with a `CacheMultiStore` as is currently done for `BaseApp` queries.

### Internal Methods

In many cases, we may wish for modules to call methods on other modules which are not exposed to clients at all. For this
purpose, we add the `InternalServer` method to `Configurator`:

```go
type Configurator interface {
   MsgServer() grpc.Server
   QueryServer() grpc.Server

>    InternalServer() grpc.Server

Reviewer 4: This means that there will be an optional `internal.proto` file per module, with a protobuf service called `Internal`, right?

I would add a paragraph to make this explicit.

}
```

> Services registered against `InternalServer` will be callable from other modules but not by external clients.

Reviewer 8: Could you list some use-cases where a module needs to call a "protected" RPC? I think it's better to keep that at minimum or avoid at all.

Author: @amaurymartiny brought up the use case of `x/staking` interacting with other modules. Not that internal messages would still follow the same authentication rules so it wouldn't be a security risk. I can add a note about that.

Reviewer 4: I believe https://github.com/cosmos/cosmos-sdk/pull/7459#discussion_r509275185 could be one example: x/slashing's Slash calls x/staking's Slash, but we don't want to expose x/staking's Slash.

### Authorization

> By default, the inter-module router requires that modules are sent by the first signer returned by `GetSigners`. The

Reviewer 8: ```suggestion
By default, the inter-module router requires that messages are sent by the first signer returned by `GetSigners`. The
```

  * optimizes inter-module calls - for instance caching resolved methods after first invocation
* combining `StoreKey`s and `ModuleKey`s into a single interface so that modules have a single Ocaps handle
* code generation which makes inter-module communication more performant
* decoupling `ModuleKey` creation from `AppModuleBasic.Name()` so that app's can override root module account names
* inter-module hooks and plugins

## Alternatives

The `x/capability` module does provide a proper object-capability implementation that can be used by any module in the
SDK and could even be used for inter-module Ocaps as described in [\#5931](https://github.com/cosmos/cosmos-sdk/issues/5931).

The advantages of the approach described in this ADR are mostly around how it integrates with other parts of the SDK,
specifically:

* protobuf so that:
  * code generation of interfaces can be leveraged for a better dev UX
  * module interfaces are versioned and checked for breakage using [buf](https://docs.buf.build/breaking-overview)
* sub-module accounts as per ADR 028
* the general `Msg` passing paradigm and the way signers are specified by `GetSigners`

> Also, this a complete replacement for keepers and could be applied to _all_ inter-module communication whereas the

Reviewer 8: ```suggestion
Also, this is a complete replacement for keepers and could be applied to _all_ inter-module communication whereas the
```

3. The router for inter-module communication gives us a convenient place to handle rollback of transactions,
enabling atomicy of operations ([currently a problem](https://github.com/cosmos/cosmos-sdk/issues/8030)). Any failure within a module-to-module call would result in a failure of the entire
transaction

This mechanism has the added benefits of:
- reducing boilerplate through code generation, and
- allowing for modules in other languages either via a VM like CosmWasm or sub-processes using gRPC 

### Inter-module Communication

To use the `Client` generated by the protobuf compiler we need a `grpc.ClientConn` implementation. For this we introduce
a new type, `ModuleKey`, which implements the `grpc.ClientConn` interface. `ModuleKey` can be thought of as the "private
key" corresponding to a module account, where authentication is provided through use of a special `Invoker()` function, 
described in more detail below.

Whereas external clients use their account's private key to sign transactions containing `Msg`s where they are listed as signers,
modules use their `ModuleKey` to send `Msg`s where they are listed as the sole signer to other modules. For example, modules

> could use their `ModuleKey` to "sign" a `/cosmos.bank.Msg/Send` transaction to send coins from the module's account to another

Author: Can you think of a different conceptual framing?

> could use their `ModuleKey` to "sign" a `/cosmos.bank.Msg/Send` transaction to send coins from the module's account to another

Reviewer 7: I think associating signatures with messages is a nice idea.

We can keep the same format as is, and just pass this info when dispatching each message (all tx signers).

The fact that this is enforced in the ante-handler (1x per tx) and not in the message handler (for each msg) would require a refactor from the current sdk design to match this design. I think that could be a good thing. It just needs to be defined

  ModuleName string
  Path []byte
}

func (key ModuleID) Address() []byte {
  return AddressHash(key.ModuleName, key.Path)
}
```

In addition to being able to generate a `ModuleID` and address, a `ModuleKey` contains a special function closure called
the `Invoker` which is the key to safe inter-module access. The `InvokeFn` closure corresponds to the `Invoke` method in
the `grpc.ClientConn` interface and under the hood is able to route messages to the appropriate `Msg` and `Query` handlers
performing appropriate security checks on `Msg`s. This allows for even safer inter-module access than keeper's whose
private member variables could be manipulated through reflection. Golang does not support reflection on a function
closure's captured variables and direct manipulation of memory would be needed for a truly malicious module to bypass
the `ModuleKey` security.

The two `ModuleKey` types are `RootModuleKey` and `DerivedModuleKey`:

```go
func Invoker(callInfo CallInfo) func(ctx context.Context, request, response interface{}, opts ...interface{}) error

type CallInfo {
  Method string
  Caller ModuleID
}

type RootModuleKey struct {
  moduleName string
  invoker Invoker

> }

Reviewer 8: ```suggestion
}

func (rm RootModuleKey) Derive(path []byte) DerivedModuleKey {}
```

Reviewer 8: Let's include the method type here.

type DerivedModuleKey struct {
  moduleName string
  path []byte

>   invoker Invoker

Reviewer 8: How about renaming `invoker Invoker` to `mkInvoke Invoker` in both structures?

}
```

A module can get access to a `DerivedModuleKey`, using the `Derive(path []byte)` method on `RootModuleKey` and then
would use this key to authenticate `Msg`s from a sub-account. Ex:

```go
package foo

func (fooMsgServer *MsgServer) Bar(ctx context.Context, req *MsgBar) (*MsgBarResponse, error) {

>   derivedKey := fooMsgServer.moduleKey.Derive(req.SomePath)

Reviewer 7: This new way of authenticating messages with a singer is "hand-waved" into existence and referenced below as an advantage:

> the general `Msg` passing paradigm and the way signers are specified by `GetSigners`

However, I did not see that explicitly specified in this doc (maybe as it is hard to read through all the comments, maybe as it is not in here). It deserves a section (or ADR) of it's own.

from either the root or a derived module account to pass through.

Note that `Invoker` itself returns a function closure based on the `CallInfo` passed in. This will allow client implementations
in the future that cache the invoke function for each method type avoiding the overhead of hash table lookup.
This would reduce the performance overhead of this inter-module communication method to the bare minimum required for
checking permissions.

To re-iterate, the closure only allows access to authorized calls. There is no access to anything else regardless of any
name impersonation.

Below is a rough sketch of the implementation of `grpc.ClientConn.Invoke` for `RootModuleKey`:

```go
func (key RootModuleKey) Invoke(ctx context.Context, method string, args, reply interface{}, opts ...grpc.CallOption) error {
  f := key.invoker(CallInfo {Method: method, Caller: ModuleID {ModuleName: key.moduleName}})
  return f(ctx, args, reply)
}
```

### `AppModule` Wiring and Requirements

In [ADR 031](./adr-031-msg-service.md), the `AppModule.RegisterService(Configurator)` method was introduced. To support

> inter-module communication, we extend the `Configurator` interface to pass in the `ModuleKey` and to allow modules to

Author: The `ModuleKey` is passed into the module implementation by the `Configurator`. It doesn't get exposed to other modules, just the module that it belongs to. Basically the `Configurator` gives you your private key and doesn't show it to anyone else.

> inter-module communication, we extend the `Configurator` interface to pass in the `ModuleKey` and to allow modules to

Reviewer 7: Ah, I didn't realize how the Configurator works. That makes sense

specify their dependencies on other modules using `RequireServer()`:

```go
type Configurator interface {
   MsgServer() grpc.Server
   QueryServer() grpc.Server

   ModuleKey() ModuleKey

>    RequireServer(serverInterface interface{})

Reviewer 8: Can we have a better interface type than empty interface? Otherwise please add a comment, that here we are expecting `MsgServer` types.

}
```

The `ModuleKey` is passed to modules in the `RegisterService` method itself so that `RegisterServices` serves as a single
entry point for configuring module services. This is intended to also have the side-effect of greatly reducing boilerplate in

> `app.go`. For now, `ModuleKey`s will be created based on `AppModuleBasic.Name()`, but a more flexible system may be

Author: It is based on a wrapped go closure so it can only be forged by manipulating the call stack. There are no hidden strings used for secure access.

introduced in the future. The `ModuleManager` will handle creation of module accounts behind the scenes.

Because modules do not get direct access to each other anymore, modules may have unfulfilled dependencies. To make sure
that module dependencies are resolved at startup, the `Configurator.RequireServer` method should be added. The `ModuleManager`
will make sure that all dependencies declared with `RequireServer` can be resolved before the app starts. An example
module `foo` could declare it's dependency on `x/bank` like this:

```go
package foo

func (am AppModule) RegisterServices(cfg Configurator) {
  cfg.RequireServer((*bank.QueryServer)(nil))
  cfg.RequireServer((*bank.MsgServer)(nil))
}
```

### Security Considerations

In addition to checking for `ModuleKey` permissions, a few additional security precautions will need to be taken by
the underlying router infrastructure.

#### Recursion and Re-entry

Recursive or re-entrant method invocations pose a potential security threat. This can be a problem if Module A
calls Module B and Module B calls module A again in the same call.

> One basic way for the router system to deal with this is to maintain a call stack which prevents a module from

Author: Yeah, I share a similar concern. Maybe re-entrancy is not a concern for modules like it is for user-defined smart contracts.

being referenced more than once in the call stack so that there is no re-entry. A `map[string]interface{}` table
in the router could be used to perform this security check.

#### Queries

Queries in Cosmos SDK are generally un-permissioned so allowing one module to query another module should not pose
any major security threats assuming basic precautions are taken. The basic precaution that the router system will
need to take is making sure that the `sdk.Context` passed to query methods does not allow writing to the store. This
can be done for now with a `CacheMultiStore` as is currently done for `BaseApp` queries.

> ### Internal Methods

Reviewer 8: ```suggestion
### Extension: Internal Methods
```

> In many cases, we may wish for modules to call methods on other modules which are not exposed to clients at all. For this

Reviewer 7: I agree with this.
There are "UserMessages" (which can be directly called from plenty of untrusted actors) and "InternalMessages" (which must be passed from a relatively trusted internal module, although with some configurable permissioning inside).

I think the threat model should be made more explicit. We assume external accounts could be hackers and are totally untrusted. What assumptions do we make on internal modules?

In CosmWasm, we assume contracts are as untrusted as external accounts. In the Cosmos SDK, they require a very high level of trust on any module as soon as you pass it a keeper from another module. 

It is important to clarify what we are protecting against, or we risk making a huge stone wall for the front door (and impact normal usability), but leave a back window open.

Author: I think the main thing I've thought of regarding internal methods is that they really aren't that useful for any external user and are really just "internal".

purpose, we add the `InternalServer` method to `Configurator`:

```go
type Configurator interface {
   MsgServer() grpc.Server
   QueryServer() grpc.Server
   InternalServer() grpc.Server
}
```

As an example, x/slashing's Slash must call x/staking's Slash, but we don't want to expose x/staking's Slash to end users
and clients.

This wound also require creating a corresponding `internal.proto` file with a protobuf service in the given module's
proto package.

Services registered against `InternalServer` will be callable from other modules but not by external clients.

An alternative solution to internal-only methods could involve hooks / plugins as discussed [here](https://github.com/cosmos/cosmos-sdk/pull/7459#issuecomment-733807753).
A more detailed evaluation of a hooks / plugin system will be addressed later in follow-ups to this ADR or as a separate
ADR.

### Authorization

By default, the inter-module router requires that messages are sent by the first signer returned by `GetSigners`. The

> inter-module router should also accept authorization middleware such as that provided by [ADR 030](https://github.com/cosmos/cosmos-sdk/pull/7105).

Reviewer 8: ```suggestion
inter-module router should also accept authorization middleware such as that provided by [ADR 030](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-030-authz-module.md).
```

> inter-module router should also accept authorization middleware such as that provided by [ADR 030](https://github.com/cosmos/cosmos-sdk/pull/7105).

Reviewer 7: Now that the PR is merged, can you reference the adr in master: https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-030-authz-module.md ?

BTW, that adr saws it was accepted 2021-11-13 (in 10 months), I think you need to swap a 1 for a 0.

other modules. This will be addressed in separate ADRs or updates to this ADR.

### Future Work

Other future improvements may include:
* custom code generation that:
  * simplifies interfaces (ex. generates code with `sdk.Context` instead of `context.Context`)
  * optimizes inter-module calls - for instance caching resolved methods after first invocation
* combining `StoreKey`s and `ModuleKey`s into a single interface so that modules have a single Ocaps handle
* code generation which makes inter-module communication more performant
* decoupling `ModuleKey` creation from `AppModuleBasic.Name()` so that app's can override root module account names
* inter-module hooks and plugins

## Alternatives

### MsgServices vs `x/capability`

The `x/capability` module does provide a proper object-capability implementation that can be used by any module in the
SDK and could even be used for inter-module Ocaps as described in [\#5931](https://github.com/cosmos/cosmos-sdk/issues/5931).

The advantages of the approach described in this ADR are mostly around how it integrates with other parts of the SDK,
specifically:

* protobuf so that:

>   * code generation of interfaces can be leveraged for a better dev UX

Reviewer 7: Is there code generation?

https://github.com/cosmos/cosmos-sdk/blob/master/x/bank/keeper/msg_server.go seems to be hand-written (and we had to hand-code the equivalent for `x/wasm`). Until this is auto-generated, I see little Dev UX benefit of the MsgServer.

Author: Well you have to hand code the implementation, but the interface is code generated. What are you hoping would be auto-generated that isn't? (We have worked on some custom code generation btw #8270).

>   * module interfaces are versioned and checked for breakage using [buf](https://docs.buf.build/breaking-overview)

Reviewer 7: This is a good point - stable interfaces.

The question is would the sdk team actually commit to no API breaking changes?

Author: That's the idea 😄
